### ISSUE TITLE
Extra metrics are being sent to DataDog when using Allow list with pattern matching enabled

### ISSUE THREAD

**Issue by rawwar** [apache/airflow#39400]
### Apache Airflow version

main (development)

### What happened?

with `AIRFLOW__METRICS__METRICS_USE_PATTERN_MATCH` set to `true` and `AIRFLOW__METRICS__METRICS_ALLOW_LIST` set to `.*get_astronauts.*` , I notice several other metrics being sent to datadog that don't match the pattern:
![image](https://github.com/apache/airflow/assets/20266953/16510682-9e45-48c1-9a95-ce726996cca5)

### What you think should happen instead?

Following metrics shouldn't be showing up on datadog when the allow list is `.*get_astronauts.*`:

1. `airflow.dag.task.duration.*`
2. `airflow.ti.*`
3. `datadog.*` (Not sure if we have any control to block these)

### How to reproduce

env variables:

AIRFLOW__METRICS__STATSD_DATADOG_ENABLED | true
-- | -- 
AIRFLOW__METRICS__STATSD_DATADOG_TAGS | testNo:C2
DATADOG_API_KEY | ● ● ● ● ● ● ●
AIRFLOW__METRICS__METRICS_USE_PATTERN_MATCH | true
AIRFLOW__METRICS__METRICS_ALLOW_LIST | `.*print_astronaut_craft.*`

Once you update your datadog key here, you will start seeing metrics in datadog-->metrics-->summary . use the tag `testNo:C2` to filter

### Operating System

Debian GNU/Linux 11 (bullseye)

### Deployment

Astronomer

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by eladkal** [comment]
@rawwar is this still an issue on main?

**Comment by rawwar** [comment]
@eladkal , Yes. i just tested this and I can still replicate it. I added a allow list and yet, I see airflow.dag.task.duration.*, airflow.ti.*

**Comment by rawwar** [comment]
However, I've unassigned myself as I won't be able to work on it at this time.

**Comment by eladkal** [comment]
cc @hussein-awala maybe someone from datadog has the time to pick it up?

**Comment by hussein-awala** [comment]
I had a look at this on main and I don't think it's still valid.

`metrics_use_pattern_match` was removed in Airflow 3 — pattern matching is the only
mode now, so the config in the repro doesn't exist anymore. And the allow list itself
is applied on every emit path in the Datadog logger (incr, decr, gauge, timing, timer),
including for legacy metric names, so nothing is slipping past the filter.

What was really happening is that `ti.start`, `ti.finish` and `task.duration` don't have
the dag_id in the metric *name* — it's in the tags. So a pattern like
`.*get_astronauts.*` was never going to match them. It only matches the legacy names
(`dag.{dag_id}.{task_id}.duration` and friends), which are behind `legacy_names_on`. The
`datadog.*` ones are the DogStatsd client's own telemetry and were always outside our
allow list.

So this isn't a filtering bug — it's that allow/block lists match on names and aren't
tag-aware. That's a reasonable thing to want, but it's a different change from what's
described here, so I'd rather see it as its own feature request. Closing this one.

---
Drafted-by: Claude Code (Opus 5); reviewed by @hussein-awala before posting
